Library/OcAfterBootCompatLib/ServiceOverrides.c

```diff
@@ -317,6 +318,117 @@ DevirtualiseMmio (
   }
 }
 
+/**
+  Apply single booter patch.
+
+  @param[in,out]  ImageBase      Booter image to be patched.
+  @param[in]      ImageSize      Size of booter image.
+  @param[in]      Patch          Single patch to be applied to booter.
+**/
+STATIC
+VOID
+ApplyBooterPatch (
+  IN OUT UINT8            *ImageBase,
+  IN     UINTN            ImageSize,
+  IN     OC_BOOTER_PATCH  *Patch
+  )
+{
+  UINT32  ReplaceCount;
+
+  if (ImageSize < Patch->Size) {
+    DEBUG ((DEBUG_INFO, "OCABC: Image size is even smaller than patch size\n"));
+    return;
+  }
+
+  if (Patch->Limit > 0 && Patch->Limit < ImageSize) {
+    ImageSize = Patch->Limit;
+  }
+
+  ReplaceCount = ApplyPatch (
+    Patch->Find,
+    Patch->Mask,
+    Patch->Size,
+    Patch->Replace,
+    Patch->ReplaceMask,
+    ImageBase,
+    (UINT32) ImageSize,
+    Patch->Count,
+    Patch->Skip
+    );
+
+  if (ReplaceCount > 0 && Patch->Count > 0 && ReplaceCount != Patch->Count) {
+    DEBUG ((
+      DEBUG_INFO,
+      "OCABC: Booter patch (%a) performed only %u replacements out of %u\n",
+      Patch->Comment,
+      ReplaceCount,
+      Patch->Count
+      ));
+  } else if (ReplaceCount == 0) {
+    DEBUG ((
+      DEBUG_INFO,
+      "OCABC: Failed to apply Booter patch (%a) - not found\n",
+      Patch->Comment
+      ));
+  } else {
+    DEBUG ((DEBUG_INFO, "OCABC: Booter patch (%a) replace count - %u\n", Patch->Comment, ReplaceCount));
+  }
+}
+
+/**
+  Iterate through user booter patches and apply them.
+**/
+STATIC
+VOID
+ApplyBooterPatches (
+  IN  EFI_HANDLE       ImageHandle,
+  IN  BOOLEAN          IsApple,
+  IN  OC_BOOTER_PATCH  *Patches,
+  IN  UINT32           PatchCount
+  )
+{
+  EFI_STATUS                  Status;
+  EFI_LOADED_IMAGE_PROTOCOL   *LoadedImage;
+  UINT32                      Index;
+  BOOLEAN                     UsePatch;
+  CONST CHAR8                 *UserIdentifier;
+  CHAR16                      *UserIdentifierUnicode;
+
+  Status = gBS->HandleProtocol (
+    ImageHandle,
+    &gEfiLoadedImageProtocolGuid,
+    (VOID **)&LoadedImage
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "OCABC: Failed to handle LoadedImage protocol - %r", Status));
+    return;
+  }
+
+  for (Index = 0; Index < PatchCount; ++Index) {
+    UserIdentifier = Patches[Index].Identifier;
+
+    UsePatch = UserIdentifier[0] == '\0' || AsciiStrCmp (UserIdentifier, "Any") == 0;
+    if (AsciiStrCmp (UserIdentifier, "Apple") == 0) {
+      UsePatch = IsApple;
+    } else {
+      UserIdentifierUnicode = AsciiStrCopyToUnicode (UserIdentifier, 0);
+      if (UserIdentifierUnicode == NULL) {
+        return;
```

Comment:
Disagree that it should not report at least something and that it should abort the whole patching process. Like this maybe?

```C
if (UserIdentifierUnicode == NULL) {
  DEBUG ((DEBUG_INFO, "OCABC: Booter patch (%a) for %a is our of memory\n", Patches[Index].Comment, UserIdentifier));
  continue;
}
```